protobuf/lib/src/protobuf/proto3_json.dart

```diff
@@ -89,11 +100,20 @@ Object? _writeToProto3Json(_FieldSet fs, TypeRegistry typeRegistry) {
   var result = <String, dynamic>{};
   for (var fieldInfo in fs._infosSortedByTag) {
     var value = fs._values[fieldInfo.index!];
-    if (value == null || (value is List && value.isEmpty)) {
-      continue; // It's missing, repeated, or an empty byte array.
-    }
     dynamic jsonValue;
-    if (fieldInfo.isMapField) {
+    if (context.emitDefaults &&
+        ((fieldInfo.isRepeated && isDefaultListField(value)) ||
+            (fieldInfo.isMapField && isDefaultMapField(value)))) {
+      jsonValue = fieldInfo.readonlyDefault;
+    } else if (context.emitDefaults &&
+        ((_isBytes(fieldInfo.type) && isDefaultListField(value)) ||
+            (_isGroupOrMessage(fieldInfo.type) && value == null))) {
+      jsonValue = null;
+    } else if (context.emitDefaults && value == null) {
+      jsonValue = valueToProto3Json(fieldInfo.readonlyDefault, fieldInfo.type);
+    } else if (isDefaultListField(value)) {
+      continue; // It's missing, repeated, or an empty byte array.
+    } else if (fieldInfo.isMapField) {
```

Comment:
> However, I just realized (as I was typing this) that you cannot use FieldSet._gitField in tests, because that function uses BuilderInfo to check the field type and return the default values depending on that, and we don't populate BuilderInfo with fields in the mock class.

Sorry, I missed this code: https://github.com/google/protobuf.dart/blob/master/protobuf/test/mock_util.dart#L16-L33

So it seems like we should be able to simplify this code by using `_getField` instead of `_values`.

Reply:
Working on implementing these changes now.

Reply:
@osa1 I took a first pass at this implementation.  Please have a look and let me know your thoughts.